File: frappe_tools/frappe_tools/report/document_upload_status_report/document_upload_status_report.js
// Copyright (c) 2026, sakthi123msd@gmail.com and contributors
// For license information, please see license.txt

const UPLOAD_REPORT_METHOD =
  "frappe_tools.frappe_tools.report.scanned_document_detail_report.scanned_document_detail_report";

// fieldname -> { field, type, default } for the currently selected document type
let UPLOAD_DATE_FIELD_CONFIG = {};

frappe.query_reports["Document Upload Status Report"] = {
  disable_auto_run: true,
  onload: function (report) {
    frappe.call({
      method: UPLOAD_REPORT_METHOD + ".get_document_types",
      callback: function (r) {
        if (r.message) {
          report.set_filter_value("document_type", null);
          report.filters[0].df.options = r.message.join("\n");
          report.filters[0].refresh();
        }
      },
    });
    // keep the date range hidden until a field is chosen
    upload_toggle_date_filters();
  },

  filters: [
    {
      fieldname: "document_type",
      label: __("Document Type"),
      fieldtype: "Select",
      reqd: 1,
      options: "",
      on_change: function () {
        upload_load_filter_fields();
      },
    },
    {
      fieldname: "filter_field",
      label: __("Filter By Field"),
      fieldtype: "Select",
      reqd: 1,
      options: "",
      on_change: function () {
        upload_toggle_date_filters();
      },
    },
    {
      fieldname: "date_start",
      label: __("Start Date"),
      fieldtype: "Date",
    },
    {
      fieldname: "date_end",
      label: __("End Date"),
      fieldtype: "Date",
    },
    {
      fieldname: "datetime_start",
      label: __("Start Date/Time"),
      fieldtype: "Datetime",
    },
    {
      fieldname: "datetime_end",
      label: __("End Date/Time"),
      fieldtype: "Datetime",
    },
    {
      fieldname: "scan_status",
      label: __("Scan Status"),
      fieldtype: "Select",
      options: ["", "Scanned", "Not Scanned"].join("\n"),
    },
  ],

  formatter: function (value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);
    if (column.fieldname === "is_scanned") {
      if (value) {
        return `
          <span class="indicator green">
            Scanned
          </span>
        `;
      }
      return `
        <span class="indicator red">
          Not Scanned
        </span>
      `;
    }
    return value;
  },
};

// Pull the configured date fields for the selected document type and
// populate the "Filter By Field" dropdown.
function upload_load_filter_fields() {
  const report = frappe.query_report;
  const document_type = report.get_filter_value("document_type");
  const field_filter = report.get_filter("filter_field");

  UPLOAD_DATE_FIELD_CONFIG = {};

  if (!document_type) {
    field_filter.df.options = "";
    field_filter.set_value("");
    field_filter.refresh();
    upload_toggle_date_filters();
    return;
  }

  frappe.call({
    method: UPLOAD_REPORT_METHOD + ".get_report_date_fields",
    args: { document_type: document_type },
    callback: function (r) {
      const fields = r.message || [];
      fields.forEach(function (f) {
        UPLOAD_DATE_FIELD_CONFIG[f.field] = f;
      });
      // show the label in the dropdown, keep the fieldname as the stored value
      field_filter.df.options = fields.map(function (f) {
        return { value: f.field, label: f.label || f.field };
      });
      // auto-select the first configured field
      field_filter.set_value(fields.length ? fields[0].field : "");
      field_filter.refresh();
      upload_toggle_date_filters();
    },
  });
}

// Show the Date or the Datetime range based on the selected field's
// configured type, hide the other, and apply any default.
function upload_toggle_date_filters() {
  const report = frappe.query_report;
  const date_start = report.get_filter("date_start");
  const date_end = report.get_filter("date_end");
  const dt_start = report.get_filter("datetime_start");
  const dt_end = report.get_filter("datetime_end");

  if (!date_start || !dt_start) return;

  const selected = report.get_filter_value("filter_field");
  const conf = UPLOAD_DATE_FIELD_CONFIG[selected] || {};
  const is_datetime = conf.type === "Datetime";
  const has_field = !!selected;

  date_start.$wrapper.toggle(has_field && !is_datetime);
  date_end.$wrapper.toggle(has_field && !is_datetime);
  dt_start.$wrapper.toggle(has_field && is_datetime);
  dt_end.$wrapper.toggle(has_field && is_datetime);

  if (has_field && conf.default) {
    const start = is_datetime ? dt_start : date_start;
    if (!start.get_value()) {
      start.set_value(conf.default);
    }
  }
}

File: frappe_tools/frappe_tools/report/document_upload_status_report/document_upload_status_report.py
import frappe

from frappe_tools.frappe_tools.report.scanned_document_detail_report.scanned_document_detail_report import (
	get_filter_field_config,
)


def execute(filters=None):
    filters = filters or {}

    doctype = filters.get("document_type")
    if not doctype:
        frappe.throw("DocType is required")

    filter_field = filters.get("filter_field")
    if not filter_field:
        frappe.throw("Please select the field to filter by")

    field_conf = get_filter_field_config(doctype, filter_field)

    if field_conf.get("type") == "Datetime":
        from_date = filters.get("datetime_start")
        to_date = filters.get("datetime_end")
    else:
        from_date = filters.get("date_start")
        to_date = filters.get("date_end")

    if not from_date or not to_date:
        frappe.throw("Please set the start and end range for the selected field")

    meta = frappe.get_meta(doctype)

    listview_fields = get_listview_fields(meta)

    columns = []
    for fieldname in listview_fields:
        df = meta.get_field(fieldname)

        columns.append({
            "label": df.label if df else fieldname.replace("_", " ").title(),
            "fieldname": fieldname,
            "fieldtype": df.fieldtype if df else "Data",
            "width": 150,
        })

    columns.append({
        "label": "Scan Status",
        "fieldname": "is_scanned",
        "fieldtype": "Data",
        "width": 200,
    })

    query_filters = {"docstatus": ["<", 2]}
    query_filters[filter_field] = ["between", [from_date, to_date]]

    data = frappe.get_all(
        doctype,
        filters=query_filters,
        fields=listview_fields,
        order_by="modified desc",
    )

    docnames = [d.name for d in data]

    scanned = frappe.get_all(
        "Scanned Document",
        filters={
            "_doctype": doctype,
            "_docname": ["in", docnames],
        },
        pluck="_docname",
    )

    scanned_set = set(scanned)

    for row in data:
        row["is_scanned"] = 1 if row.name in scanned_set else 0

    scan_status = filters.get("scan_status")
    if scan_status == "Scanned":
        data = [row for row in data if row.is_scanned]
    elif scan_status == "Not Scanned":
        data = [row for row in data if not row.is_scanned]

    return columns, data

def get_listview_fields(meta):
    fields = []

    if meta.title_field:
        fields.append(meta.title_field)

    for df in meta.fields:
        if df.in_list_view and df.fieldname not in fields:
            fields.append(df.fieldname)

    if "name" not in fields:
        fields.insert(0, "name")

    if "modified" not in fields:
        fields.append("modified")

    return fields
